vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -295,17 +297,37 @@ def process_weights_after_loading(self, layer: Module) -> None:
         # If the kv-cache dtype is auto, we enforce the kv-scale to be 1.0
         # regardless whether the kv-scale is available in the checkpoint.
         if layer.kv_cache_dtype != "auto":
-            kv_scale = layer.kv_scale.to("cpu").tolist()
-            if not isinstance(kv_scale, float):
-                raise ValueError("Only support per-tensor scaling factor "
-                                 "for fp8 KV cache")
-            layer._kv_scale = kv_scale
-            if layer._kv_scale == 1.0 and "e5m2" not in layer.kv_cache_dtype:
-                print_warning_once(
-                    "Using KV cache scaling factor 1.0 for fp8_e4m3. This may "
-                    "cause accuracy issues. Please make sure kv-cache scaling "
-                    "factor is available in the fp8 checkpoint.")
+            # We prefer to use separate key_scale and value_scale if present
+            if layer.key_scale > 0.0 and layer.value_scale > 0.0:
+                key_scale = layer.key_scale.to("cpu").tolist()
+                value_scale = layer.value_scale.to("cpu").tolist()
+                if not isinstance(key_scale, float) or not isinstance(
+                        value_scale, float):
+                    raise ValueError("Only support per-tensor scaling factor "
+                                     "for fp8 KV cache")
+                layer._key_scale = key_scale
+                layer._value_scale = value_scale
+                if (layer._key_scale == 1.0 and layer._value_scale == 1.0
+                        and "e5m2" not in layer.kv_cache_dtype):
+                    print_warning_once(
+                        "Using KV cache scaling factor 1.0 for fp8_e4m3. This "
+                        "may cause accuracy issues. Please make sure kv-cache "
+                        "scaling factor is available in the fp8 checkpoint.")
+            else:
+                kv_scale = layer.kv_scale.to("cpu").tolist()
+                if not isinstance(kv_scale, float):
+                    raise ValueError("Only support per-tensor scaling factor "
+                                     "for fp8 KV cache")
+                layer._kv_scale = kv_scale
+                if (layer._kv_scale == 1.0
+                        and "e5m2" not in layer.kv_cache_dtype):
+                    print_warning_once(
+                        "Using KV cache scaling factor 1.0 for fp8_e4m3. This "
+                        "may cause accuracy issues. Please make sure kv-cache "
+                        "scaling factor is available in the fp8 checkpoint.")
```

Comment:
IIUC this is for back compatibility? If so I'd suggest make it explicitly, even with deprecate warning.

Another way I prefer more is only maintaining separated scales in model parameters. For example, we could remap kv_scale to key_scale, and and duplicate key_scale to value_scale here if we found that value_scale is 0, which is invalid anyways.

Reply:
Yes, I have light deprecation message elsewhere but will make it more explicit. I like the idea of just mapping kv_scale -> key_scale, will implement!